pkg/highlight/highlighter.go

```diff
@@ -274,20 +280,28 @@ func (h *Highlighter) HighlightString(input string) []LineMatch {
 	for i := 0; i < len(lines); i++ {
 		line := []byte(lines[i])
 		highlights := make(LineMatch)
-		lineMatches = append(lineMatches, h.highlight(highlights, 0, i, line, nil))
+		lineMatches = append(lineMatches, h.highlight(highlights, 0, i, line, nil, false))
 	}
 
 	return lineMatches
 }
 
 // Highlight sets the state and matches for each line from startline to endline
 // It sets all other matches in the buffer to nil to conserve memory
-func (h *Highlighter) Highlight(input LineStates, startline, endline int) {
+// In the moment it is called with statesOnly set it will scan down from
+// `startline` and set the appropriate end of line state for each line
+// until it comes across a line whose state does not change
+// returns the number of the final line
+func (h *Highlighter) Highlight(input LineStates, startline, endline int, statesOnly bool) int {
```

Comment:
To make the API more intuitive, I'd suggest to remove `statesOnly` from `Highlight()` (but not from the internal `highlight()` function) and restore the `ReHighlightStates()` function which would call `highlight()` with statesOnly=true.

P.S. This is not the most important thing at the moment. I'd suggest to first address the regression reported by @niten94 in https://github.com/zyedidia/micro/pull/3127#issuecomment-1937714385 because maybe it means that the entire approach of this PR is wrong. (I haven't thoroughly reviewed this PR yet.)